sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java

```diff
@@ -194,14 +202,41 @@ public void record(Attributes attributes, T accumulation) {
                 + ").");
         return;
       }
-      // TODO: error on metric overwrites
+
+      // Check there is not already a recording for the attributes
+      if (currentAccumulation.containsKey(attributes)) {
+        logger.log(
+            Level.WARNING,
+            "Instrument "
+                + instrument.getName()
+                + " has recorded multiple values for the same attributes.");
+        return;
+      }
+
+      // Check that the accumulation is compatible with the previously recorded accumulation
+      // e.g. ensure that monotonic sum values are monotonically increasing
+      T previous = previousAccumulation.get(attributes);
+      if (previous != null) {
+        try {
+          aggregator.validateAsyncAccumulation(previous, accumulation);
+        } catch (Exception e) {
+          // TODO: consider resetting the metric stream instead of logging here
+          logger.log(
+              Level.WARNING,
+              "Instrument "
+                  + instrument.getName()
+                  + " has recorded an invalid value: "
+                  + e.getMessage());
+          return;
+        }
+      }
```

Comment:
@jsuereth curious what you think about this. Repeating a blurb from the PR description for context:

Observable sums must be monotonically increasing. Current behavior is to allow it. New behavior is to log with a throttling logger, if and only if the default sum aggregation is used.

Synchronous instruments don't need access to previous recordings in order to ensure the rules are followed, but observable sums do since you need to compare the current recorded value to the previous. My implementation adds this check to the AsynchronousMetricStorage, which retains a copy of previous accumulations, and asks the configured Aggregator if the current accumulation is valid given the previous. The problem with this is that the check is ignored if you switch the aggregator to something else, such as last value. But there isn't any other natural place to store the previous value such that a check can be consistently implemented.

Reply:
Do you still report the value?

It is allowed, and acceptable, for cumulative sum to "roll" i.e.. go past int max and return int min.  These are commonly used to compute rates at query time, so they're always diffed and backends generally have wrap-around protection logic.

While I understand what you want to do here, I think we could allow users to provide any value and document expectations.  We shouldn't limit async in this way unless we're positive we're not preventing legitimate uses (which is almost impossible)

Reply:
Yeah I mentioned in the PR description but I'll say it again, I'm very unsure about the value of this check. I found a situation where you can use the API / SDK to produce OTLP data that violates the monotonicity, which seems wrong. Made the change to prompt conversation so we don't accidentally release something we regret (FYI, this only affects the SDK, not the API). 

My main problem is that the check only actually works if you're using an async counter and accumulating with the default sum aggregator. So if you chance to a different aggregator, you can suddenly record non-monotonic measurements. 

I don't see any obvious resolution to this. One thought I had is that maybe in certain edge cases its ok that SDKs produce OTLP data that violate rules, and its just up to backends to detect this and react. (They could either reject the data or reset the stream, etc)

Reply:
The primary issue I have with the check is it makes an important metrics use-case broken.  Specifically counter-roll-over.   Unless you can verify the counter drop is not the result of int overflow, we should not be doing this check.

Reply:
Replied [here](https://github.com/open-telemetry/opentelemetry-java/pull/3977#discussion_r771641874).

TL;DR is that I think we should remove this check. By doing so we are accepting that in certain edge cases the SDK will produce data that violates rules.